test/integration/puppetlabs/services/jruby/jruby_locking_test.clj

```diff
@@ -89,24 +96,34 @@
             (is (= [:instance-requested :instance-borrowed :instance-returned
                     :lock-requested :lock-acquired :lock-released] @events))))))))
 
-(deftest ^:integration borrow-and-return-affect-read-lock-test
-  (tk-bootstrap/with-app-with-config
-    app
-    [jruby-service/jruby-puppet-pooled-service
-     profiler/puppet-profiler-service]
-    (jruby-service-test-config 1)
-    (let [jruby-service (tk-app/get-service app :JRubyPuppetService)
-          lock (jruby-testutils/get-lock-from-app app)]
-
-      (is (= 0 (.getReadLockCount lock)))
-      (is (not (.isWriteLocked lock)))
-
-      (let [instance (jruby-protocol/borrow-instance jruby-service :test-borrow-and-read-lock)]
-        (testing "borrow-instance acquires a read lock and does not affect write lock"
-          (is (= 1 (.getReadLockCount lock)))
-          (is (not (.isWriteLocked lock))))
-
-        (testing "return-instance releases a read lock and does not affect write lock"
-          (jruby-protocol/return-instance jruby-service instance :test-return-and-read-lock)
-          (is (= 0 (.getReadLockCount lock)))
-          (is (not (.isWriteLocked lock))))))))
+(deftest ^:integration with-lock-and-borrow-contention-test
+  (testing "contention for instances with borrows and locking handled properly"
+    (tk-bootstrap/with-app-with-config
+     app
+     [jruby-service/jruby-puppet-pooled-service
+      profiler/puppet-profiler-service]
+     (jruby-service-test-config 2)
+     (jruby-testutils/wait-for-jrubies app)
+     (let [jruby-service (tk-app/get-service app :JRubyPuppetService)]
+       (let [instance (jruby-protocol/borrow-instance
+                       jruby-service
+                       :with-lock-and-borrow-contention-test)]
+         (let [lock-thread-started? (promise)
+               unlock-thread? (promise)
+               lock-thread (future (jruby-service/with-lock
+                                    jruby-service
+                                    :with-lock-and-borrow-contention-test
+                                    (deliver lock-thread-started? true)
+                                    @unlock-thread?))]
+           (testing "lock not granted yet when instance still borrowed"
+             (is (not (realized?
+                       lock-thread-started?))))
+           (jruby-protocol/return-instance
+            jruby-service instance :with-lock-and-borrow-contention-test)
+           @lock-thread-started?
+           (testing "cannot borrow from non-locking thread when locked"
+             (is (not (can-borrow-from-different-thread? jruby-service))))
```

Comment:
It doesn't make any specific assumptions about the timeout other than that there is one.  The `borrow-timeout` is configured to 1 millisecond in this case, which allows the test to run quickly, but it seems like this test would work fine with a much larger timeout.  We know that there are 2 JRubies in the pool by this point because of the `(jruby-testutils/wait-for-jrubies app)` call earlier on, which loops until all of the expected instances are available in the pool.